src/engine.ts

```diff
@@ -399,6 +435,18 @@ export function initializePolyfill() {
       const maybeComputeState: (
         parentState: LayoutState
       ) => [LayoutState, boolean] = parentState => {
+        if (node instanceof ShadowRoot) {
```

Comment:
We shouldn't mutate the `parentState`. Instead, we should apply the conditions while computing the state for the `ShadowRoot`.

For now, we can do that by passing a `parentState => QueryDescriptorEntry[]` function to `createStateComputer`. For `ShadowRoot`, that will use the `ShadowRoot.styleSheets` and `ShadowRoot.adoptedStyleSheets` (if present) lists.

Reply:
I have all of the other change requests done, but this is the last piece. I am struggling to figure out how to pass the styles from the `ShadowRoot.adoptedStyleSheets` into `createStateComputer` (since it uses `window.getComputedStyle(element)` which doesn't work for `asoptedStyleSheets`)... could you send me some sample syntax on implementing this bit?

Reply:
We don't need to pass the style information to `createStateComputer` -- we just need the list of stylesheets, so that we can call `getDescriptorsForStyleSheet` on them.

```js
interface QueryDescriptorProvider {
  (parentState: LayoutState): ContainerConditionEntry[];
}

function createStateComputer(el: Element, queryDescriptorProvider: QueryDescriptorProvider) {
  /* ... */
}
```

When creating the `createStateComputer` for the Shadow DOM, we would pass one in that reads from the adopted styles. For everything else, we would just read from `parentState`.